### PR TITLE
DS-707: Location Filters

### PR DESCRIPTION
**Pull request by sherakama**
# READY FOR REVIEW

# Summary
- Location Facets for State/Prov and Country 
- Geospatial city search with autocomplete
- New API endpoints for address and location lookup
- No mobile styles or filters yet.

# Review By (Date)
- July 31

# Review Tasks

Review Preview Build: https://deploy-preview-879--stanford-alumni.netlify.app/events/discovery/

1. Navigate to the `/events/discovery` page
2. Expand the location facet group
3. Click on each of the three tabs (City, State/Prov, Country)
4. On the Country tab, click in to the country field
5. Validate that there are options for country
6. For bonus points, validate that the options for country match the options from the search index
7. Select an option
8. Validate that the results return the correct results
9. Validate that a chip is created with the correct name
10. Validate that the `(1)` shows up in the tab name
11. Click into the field or hover over it and click the `X` to clear the filter
12. Validate that the search returns a full set of results and that the chip is cleared and the tab label no longer shows the `(1)`
13. Click back in to the country field and select another option
14. Validate that the URL changes to match the selection
15. Click on the chip to clear the selection
16. Validate that the URL has the country parameter removed
17. Click back into the country field and select another option
18. Copy the URL
19. Close your browser or open a new tab
20. Paste and navigate to the deep link to validate pre-selection by results, chips, labels, and field values
21. Click on the city tab and validate that there is a message to clear the country search
22. Click on the state/prov tab and validate that there is a message to clear the country search
23. Click on the chip in the clear message and validate that the country filter is cleared by chips, URL, and results
24. Repeat steps 4 - 23 for the State/Prov field
25. Clear out all refinements and click on the city tab
26. Click into the city field and select `Current Location`
27. Accept the browser prompt to allow location
28. Validate that the current location option is selected in the field, there is a chip with both the distance and place, the results are returning (if you're near one), and that the URL has changed
29. Validate the deep link by copy and pasting the URL into an incognito page (note you wont be prompted to allow for the current location as it will be whatever the lat/lng was in the URL)
30. Close the incognito page
31. Click into the city field and start typing out 'Los Angeles'
32. When you see the option, select it
33. You should see a chip, URL change, the radius options enabled, and the results updated (about 19 at this point in time)
34. Click on the 25 mile radius option and see the chip, URL, and results update (about 12 results)
35. Copy the URL into an incognito browser and validate that it matches the expected state
36. In that new incognito browser click into the city field and clear it out with the `X`
37. Click back into the field and select the `current location` option
38. Block/Deny the browser request
39. Validate that you see an error message and a red line on the field
40. Close the incognito browser and navigate back to a clean `/events/discovery` page
41. Expand the online options facet
42. Make a few selections and validate that the chips, URL, and results update
43. Do the same with the United States, and International facet options
44. Finally, go to town and combine a bunch of the filters and facets to see if you get the expected results or can produce errors

# Notes
- If we can put the Canadian Province data into the `state` facet it should show up for this UI
- No mobile yet
- I'll have to catch up with your recent facet changes

### CODE REVIEW

File: netlify.toml [hunk 1]
```diff
@@ -19,7 +19,7 @@
     X-Content-Type-Options = "nosniff"
     Referrer-Policy = "strict-origin-when-cross-origin"
     Strict-Transport-Security = "max-age=31536000"
-    Permissions-Policy = "vibrate=(), geolocation=(), midi=(), notifications=(), push=(), microphone=(), camera=(), magnetometer=(), gyroscope=(), speaker=()"
```
Comment (sherakama) [inline netlify.toml:22]: We want the GeoLocation functionality now.

File: src/api/location/autocomplete.js
Comment (sherakama) [inline src/api/location/autocomplete.js:1]: New API endpoint for autocomplete suggestions. Currently nothing validating or throttling this endpoint and should probably throw a rate limit on it to avoid abuse.

File: src/api/location/lookup.js
Comment (sherakama) [inline src/api/location/lookup.js:1]: New lat/lng lookup by address string. Currently no authentication or validation so probably should add a rate limit to this at some point.

File: src/components/events-discovery/components/Facets/FacetCtx.jsx [hunk 2]
```diff
@@ -6,6 +6,8 @@ const knownFacets = [
   { attribute: 'experience', label: 'Experience' },
   { attribute: 'subject', label: 'Subject' },
   { attribute: 'location', label: 'Location' },
+  { attribute: 'usRegion', label: 'United States' },
+  { attribute: 'intRegion', label: 'International' },
 ];
```
Comment (sherakama) [inline src/components/events-discovery/components/Facets/FacetCtx.jsx:11]: I've now added these into the sub-facets.

File: src/components/events-discovery/components/Hit/Hit.jsx [hunk 3]
```diff
@@ -27,6 +27,34 @@ export const Hit = ({ hit }) => {
...
+    );
+  }
+
+  let exp = experience;
+  if (typeof experience === 'string') {
+    exp = [experience];
+  }
+
+  // END TODO
+  // ----------------------------------------------------------------------------------------------------
```
Comment (sherakama) [inline src/components/events-discovery/components/Hit/Hit.jsx:56]: These are all temporary patches to prevent the Event component from crashpants. These will need to be removed and some validation in the Event component would be good to have.

File: src/components/events-discovery/components/LocationFilter/FacetSubFilter.jsx
Comment (sherakama) [inline src/components/events-discovery/components/LocationFilter/FacetSubFilter.jsx:1]: Converted this one into a generic sub-filter.

File: src/components/events-discovery/components/LocationFilter/LocationFacetProvider.jsx
Comment (sherakama) [inline src/components/events-discovery/components/LocationFilter/LocationFacetProvider.jsx:1]: I set this up as I thought I would have used it more. Keeping it around for a touch to see if I end up needing it when I clean up the rest.

File: src/components/events-discovery/components/LocationFilter/LocationFacetTabs.jsx
Comment (sherakama) [inline src/components/events-discovery/components/LocationFilter/LocationFacetTabs.jsx:1]: New tab component for the three different location search options.

File: src/components/events-discovery/components/LocationFilter/LocationFilterClearContent.jsx
Comment (sherakama) [inline src/components/events-discovery/components/LocationFilter/LocationFilterClearContent.jsx:1]: This is the content that shows up when someone makes a choice in one of the three tabs and then moves to another one while the choice is still active. This still needs styling.

File: src/components/events-discovery/components/LocationFilter/LocationTabPanelCity.jsx [hunk 4]
```diff
@@ -0,0 +1,248 @@
...
+
+  // Is loading suggestions.
+  const [locationIsLoading, setLocationIsLoading] = useState(false);
+  // Suggestions for the location search.
+  const [locationSuggestions, setLocationSuggestions] = useState([
+    locationName || 'Current location',
+  ]);
+
+  // Handle the input typing into the location search.
+  const onInputType = async (e, query, reason) => {
```
Comment (sherakama) [inline src/components/events-discovery/components/LocationFilter/LocationTabPanelCity.jsx:45]: Function for handling the autocomplete lookup.

File: src/components/events-discovery/components/LocationFilter/LocationTabPanelCity.jsx [hunk 5]
```diff
@@ -0,0 +1,248 @@
...
+
+  // Custom Connector Hook.
+  const geo = useRadialGeoSearch();
+  const {
+    refine,
+    clearRefinements: clearGeoRefinement,
+    name: locationName,
+    radius,
+    setRadius,
+  } = geo;
```
Comment (sherakama) [inline src/components/events-discovery/components/LocationFilter/LocationTabPanelCity.jsx:35]: New custom connector plugin to handle the radial geo search as for some reason Algolia's component default to a map widget and a bounding box.

File: src/components/events-discovery/components/LocationFilter/LocationTabPanelCity.jsx [hunk 6]
```diff
@@ -0,0 +1,248 @@
...
+        break;
+      }
+      default:
+        // eslint-disable-next-line no-console
+        console.debug('Unhandled reason', reason);
+    }
+  };
+
+  // Handle the city change.
+  const onCityChange = async (e, value, reason) => {
```
Comment (sherakama) [inline src/components/events-discovery/components/LocationFilter/LocationTabPanelCity.jsx:84]: Handle the user selecting an option from the autocomplete. This needs all the error handling still.

File: src/components/events-discovery/components/LocationFilter/LocationTabPanelCity.jsx [hunk 7]
```diff
@@ -0,0 +1,248 @@
...
+        clearGeoRefinement();
+        break;
+      }
+      default:
+        console.log('Unknown reason', reason);
+    }
+  };
+
+  // If not active. Return.
+  if (activeTab !== 'city') return null;
```
Comment (sherakama) [inline src/components/events-discovery/components/LocationFilter/LocationTabPanelCity.jsx:136]: Don't render this panel tab if it is not active.

File: src/components/events-discovery/components/LocationFilter/LocationTabPanelCity.jsx [hunk 8]
```diff
@@ -0,0 +1,248 @@
...
+
```
Comment (sherakama) [inline src/components/events-discovery/components/LocationFilter/LocationTabPanelCity.jsx:137]: Todo: Add clear content component to this panel tab too.

File: src/components/events-discovery/components/LocationFilter/LocationTabPanelState.jsx [hunk 9]
```diff
@@ -0,0 +1,140 @@
...
+import * as styles from './LocationFilter.styles';
+import HeroIcon from '../../../simple/heroIcon';
+import LocationFilterClearContent from './LocationFilterClearContent';
+import useRadialGeoSearch from './useRadialGeoSearch';
+
+const LocationTabPanelState = () => {
+  const { activeTab } = useContext(LocationContext);
+  const searchFieldId = useId();
+  const title = 'Find a US State / Canadian Province';
+  const field = 'state';
```
Comment (sherakama) [inline src/components/events-discovery/components/LocationFilter/LocationTabPanelState.jsx:20]: I've locked this filter on to the state attribute for now. When we get around to adding or getting the province data we can change this.

File: src/components/events-discovery/components/LocationFilter/RadialGeoSearchConnector.js
Comment (sherakama) [inline src/components/events-discovery/components/LocationFilter/RadialGeoSearchConnector.js:1]: This is the new custom connector for handling the radial search requests.

This thing is the meat of the city search. It handles the adding of the three attributes to the search request and works with the URL params and UI state. 

The three params to make this search work are
1. `aroundLatLng` - Coordinates to search by a circle from
2. `aroundRadius` - How big a circle
3. `aroundPrecision`. - Variance of distance in which to consider results equal. 

This plugin provides the uiState information for the router and the renderState data for the transactions.

File: src/components/events-discovery/components/LocationFilter/RadialGeoSearchConnector.js [hunk 10]
```diff
@@ -0,0 +1,201 @@
...
+    // A `connectorState` object is used to store information
+    // that needs to be shared across multiple method calls.
+    const connectorState = {
+      radius,
+      precision,
+      name: null,
+      lat: null,
+      lng: null,
+      // Random ID to avoid collisions with other connectors.
+      id: Math.random().toString(36).substr(2, 9),
```
Comment (sherakama) [inline src/components/events-discovery/components/LocationFilter/RadialGeoSearchConnector.js:23]: TODO: Remove.

File: src/components/events-discovery/eventsDiscovery.jsx [hunk 11]
```diff
@@ -69,33 +69,52 @@ const EventsDiscovery = () => (
     future={{ preserveSharedStateOnUnmount: true }}
     stalledSearchDelay={2000}
     routing={{
-      router: history(),
+      router: history({
+        cleanUrlOnDispose: true,
```
Comment (sherakama) [inline src/components/events-discovery/eventsDiscovery.jsx:73]: Fixes a warning by setting this.

File: src/components/events-discovery/eventsDiscovery.jsx [hunk 12]
```diff
@@ -69,33 +69,52 @@ const EventsDiscovery = () => (
...
+      }),
       stateMapping: {
         stateToRoute(uiState) {
           const indexUiState = uiState['dev_alumni-events_start-asc'];
 
-          return {
-            q: indexUiState.query,
```
Comment (sherakama) [inline src/components/events-discovery/eventsDiscovery.jsx:78]: We don't have a straight search bar.

**Comment by sherakama** [comment]
@ericbakenhus 

This is as far as I am going to be able to take it this week. It's... rough but has some basis of functionality. There is a lot of state handling, UI styling, error handing, and bug fixings to do in this one still. However, you can use it to search for something and filter by state and country now.

https://deploy-preview-879--stanford-alumni.netlify.app/events/discovery/

I can pick this up again when I get back if you don't feel like fighting this battle as there is plenty of other stuff to do. 

I left some comments in the code and on this PR to help if you feel like picking it up. If you do pick it up, nothing is sacred and you can change whatever you'd like. 

Good luck.

**Comment by sherakama** [comment]
hmm, the preview build keeps failing to build all the content. Going to have to run this on your local until we get a chance to look at that.

File: src/components/events-discovery/components/Chips/ChipsComponent.jsx [hunk 13]
```diff
@@ -96,6 +105,13 @@ export const ChipsComponent = () => {
             remove={refinement.remove}
           />
         ))}
+        {LocationName && (
+          <Chip
+            attribute="city"
+            label={`Within ${miles} miles of ${LocationName}`}
+            remove={clearGeoRefinement}
+          />
+        )}
```
Comment (sherakama) [inline src/components/events-discovery/components/Chips/ChipsComponent.jsx:114]: I just stuck this at the end of the list. It is long-ish so that seemed like a good place for it to me.

File: src/components/events-discovery/components/Event/Event.jsx [hunk 14]
```diff
@@ -51,10 +51,6 @@ export const Event = ({
   experience,
   format,
   hit,
-  isDark = false,
-  isMinimal = false,
-  isBigHeadline = false,
-  headingLevel = 3,
```
Comment (sherakama) [inline src/components/events-discovery/components/Event/Event.jsx:57]: These were unused

File: src/components/events-discovery/components/FilterAccordion/FilterAccordion.jsx [hunk 15]
```diff
@@ -76,8 +76,8 @@ export const FilterAccordion = ({
       <div
         id={uniqueId}
         aria-hidden={!expanded}
-        className={dcnb('su-block su-overflow-hidden', {
-          'su-invisible': !expanded,
```
Comment (sherakama) [inline src/components/events-discovery/components/FilterAccordion/FilterAccordion.jsx:80]: I had to change these due to the field drop-down being cut off at this boundary when the overflow is hidden.

File: src/components/events-discovery/components/Hit/Hit.jsx [hunk 16]
```diff
@@ -28,6 +28,7 @@ export const Hit = ({ hit }) => {
     identityTags,
     experience,
     format,
+    subject,
```
Comment (sherakama) [inline src/components/events-discovery/components/Hit/Hit.jsx:31]: Does this need to be here? I'll need to check out the latest you have in the integration branch.

File: src/components/events-discovery/components/LocationFilter/LocationFacetTab.jsx [hunk 17]
```diff
@@ -0,0 +1,35 @@
...
+    aria-selected={isActive}
+    aria-controls={`${value}-panel`}
+    data-test={`location-${value}-tab`}
+    className={dcnb(
+      'su-text-16 su-rs-px-neg2 su-rs-py-neg2 su-text-cardinal-red-dark hocus-visible:su-bg-carinal-red-light-05 hocus:su-decoration-cardinal-red-dark su-transition',
+      {
+        'su-border-b-4 su-border-cardinal-red-light su-font-semibold': isActive,
+      },
+      { 'su-col-span-3': value !== 'state' },
+      { 'su-col-span-6': value === 'state' }
```
Comment (sherakama) [inline src/components/events-discovery/components/LocationFilter/LocationFacetTab.jsx:27]: Big words need moar space

File: src/components/events-discovery/components/LocationFilter/LocationFilter.jsx [hunk 18]
```diff
@@ -17,21 +22,28 @@ import { InternationalSubfilter } from './InternationalSubfilter';
...
+        <LocationTabPanelCity />
+        <LocationTabPanelState />
+        <LocationTabPanelCountry />
+      </LocationFacetProvider>
+
       {onlineOptionsFacet && (
-        <OnlineOptionsSubfilter
+        <FacetSubFilter
+          attribute="experience"
+          label="Online Options"
```
Comment (sherakama) [inline src/components/events-discovery/components/LocationFilter/LocationFilter.jsx:46]: This is one of the conflicts I'll need to resolve from the integration branch.

File: src/components/events-discovery/components/LocationFilter/LocationTabPanelCity.jsx [hunk 19]
```diff
@@ -0,0 +1,303 @@
...
+  });
+
+  // Custom Connector Hook.
+  const {
+    refine,
+    clearRefinements: clearGeoRefinement,
+    name: locationName,
+    radius,
+    setRadius,
+  } = useRadialGeoSearch({ primary: true });
```
Comment (sherakama) [inline src/components/events-discovery/components/LocationFilter/LocationTabPanelCity.jsx:34]: I'm not super fond of this pattern but it's good for now as multiple instances of the connector track their own state this flag tells the group of them which is the primary state to use.

File: src/components/events-discovery/components/LocationFilter/LocationTabPanelState.jsx [hunk 20]
```diff
@@ -0,0 +1,144 @@
...
+  const { items: countryItems } = useCurrentRefinements({
+    includedAttributes: ['country'],
+  });
+  const { refine: clearRefinement } = useClearRefinements({
+    includedAttributes: [field, 'country'],
+  });
+  const { items, refine } = useRefinementList({
+    attribute: field,
+    limit: 1000,
+    sortBy: ['name:asc'],
```
Comment (sherakama) [inline src/components/events-discovery/components/LocationFilter/LocationTabPanelState.jsx:34]: If we put the CDN options in to the state field we probably will want to group them a little differently and will need to change this.

File: src/components/events-discovery/components/StatusHeader/StatusHeader.jsx [hunk 21]
```diff
@@ -6,11 +6,17 @@ import { useStats } from 'react-instantsearch';
...
  */
 export const StatusHeader = () => {
-  const { nbHits, hitsPerPage, nbPages, page } = useStats();
+  const { nbHits, hitsPerPage, nbPages, page, areHitsSorted, nbSortedHits } =
+    useStats();
+
+  const numHits = useMemo(
+    () => (areHitsSorted ? nbSortedHits : nbHits),
+    [areHitsSorted, nbHits, nbSortedHits]
+  );
```
Comment (sherakama) [inline src/components/events-discovery/components/StatusHeader/StatusHeader.jsx:15]: I changed this due to the sorting issue we troubleshooter together last Tuesday. When the results are limited due to the sort, this picks that up and gives the correct number.

File: src/hooks/RadialGeoSearchConnector.js [hunk 22]
```diff
@@ -0,0 +1,249 @@
...
+        }
+
+        return connectorState;
+      },
+      /**
+       * Called once before the first search.
+       */
+      init(initOptions) {
+        const { helper } = initOptions;
+        helper.setQueryParameter('getRankingInfo', 'true');
```
Comment (sherakama) [inline src/hooks/RadialGeoSearchConnector.js:220]: Todo: remove this.

File: src/components/events-discovery/components/Hit/Hit.jsx (hunk 16 above)
Comment (ericbakenhus) [reply src/components/events-discovery/components/Hit/Hit.jsx:31]: #894 🙃

File: src/components/events-discovery/components/FilterAccordion/FilterAccordion.jsx (hunk 15 above)
Comment (ericbakenhus) [reply src/components/events-discovery/components/FilterAccordion/FilterAccordion.jsx:80]: This removes all accordion slide transitions. Are there other options? Do we care enough about the focus order weirdness caused by portals to not enable one here? Could we maybe adjust the drop-down max-height so it starts scrolling sooner and doesn't overflow the accordion?

File: src/components/events-discovery/components/LocationFilter/LocationFacetTabs.jsx [hunk 23]
```diff
@@ -0,0 +1,59 @@
...
+  const activeTabFromRefinements = items.find(
+    (item) => item.attribute === 'state' || item.attribute === 'country'
+  )?.attribute;
+
+  return (
+    <div
+      className="su-grid su-grid-cols-12 su-rs-mb-0 su-rs-mt-1 xl:su-mt-0 su-border-b su-border-black-20 xl:su-mx-0"
+      role="tablist"
+      aria-label="Location Tabs"
+    >
```
Comment (ericbakenhus) [inline src/components/events-discovery/components/LocationFilter/LocationFacetTabs.jsx:33]: It doesn't seem like we've implemented the full [tab pattern](https://www.w3.org/WAI/ARIA/apg/patterns/tabs/). I'm guessing directory might need an update as well. My Account has an example implementation (the WGH tabs).

File: src/components/events-discovery/components/LocationFilter/LocationTabPanelState.jsx (hunk 9 above)
Comment (ericbakenhus) [reply src/components/events-discovery/components/LocationFilter/LocationTabPanelState.jsx:21]: I can have the importer re-use the state attribute when adding provinces. Unless you'd think it would be clearer to have a separate province field. 🤔

File: src/components/events-discovery/components/FilterAccordion/FilterAccordion.jsx (hunk 15 above)
Comment (sherakama) [reply src/components/events-discovery/components/FilterAccordion/FilterAccordion.jsx:80]: I like the larger drop-down... Let me see what fits. Alternatively, I can make it relative and push the items below down on expand. It's a little janky but not completely noticeable.

File: src/components/events-discovery/components/Hit/Hit.jsx (hunk 16 above)
Comment (sherakama) [reply src/components/events-discovery/components/Hit/Hit.jsx:31]: Oi. Thanks.

File: src/components/events-discovery/components/FilterAccordion/FilterAccordion.jsx (hunk 15 above)
Comment (ericbakenhus) [reply src/components/events-discovery/components/FilterAccordion/FilterAccordion.jsx:80]: Coolios. Have we tried removing `disablePortal` so it's not a child element anymore? We keep it on so the element is in the "correct" DOM location, but MUI handles focus order well enough I think it's Good Enough™ in this case.

File: src/components/events-discovery/components/LocationFilter/LocationTabPanelState.jsx (hunk 9 above)
Comment (sherakama) [reply src/components/events-discovery/components/LocationFilter/LocationTabPanelState.jsx:21]: I'd be fine with slapping the data into the state field. I'd prefer to have all the data in the one field instead of having to handle two separate fields in one element.

File: src/components/events-discovery/components/FilterAccordion/FilterAccordion.jsx (hunk 15 above)
Comment (sherakama) [reply src/components/events-discovery/components/FilterAccordion/FilterAccordion.jsx]: Throwing the portal in slams the markup to the bottom of the dom and I haven't tried adding in the MUI CSS to bring it back up to the element. I can look at that too.

Comment (sherakama) [reply src/components/events-discovery/components/FilterAccordion/FilterAccordion.jsx]: Hmm that looks better now. Must have been some legacy styles that were messing with it.

File: src/components/events-discovery/components/LocationFilter/LocationFacetTabs.jsx (hunk 23 above)
Comment (sherakama) [reply src/components/events-discovery/components/LocationFilter/LocationFacetTabs.jsx]: Ok, I've added in the focus and keyboard controls that were missing. Anything else you see?

File: src/components/events-discovery/components/Facets/FacetComponent.jsx
Comment (sherakama) [inline src/components/events-discovery/components/Facets/FacetComponent.jsx:1]: Is this file still required?

Comment (sherakama) [reply src/components/events-discovery/components/Facets/FacetComponent.jsx:1]: Nope.

File: src/components/events-discovery/components/LocationFilter/LocationFacetTabs.jsx [hunk 24]
```diff
@@ -0,0 +1,110 @@
...
+          setActiveTab('country');
+          countryRef.current.focus();
+        }
+        break;
+      }
+
+      default:
+        break;
+    }
+  };
```
Comment (sherakama) [inline src/components/events-discovery/components/LocationFilter/LocationFacetTabs.jsx:72]: A little verbose and not easily expandable but I don't see these changing much if any.

**Comment by sherakama** [comment]
Ready for another pass.

File: src/components/events-discovery/components/Hit/Hit.jsx [hunk 25]
```diff
@@ -35,16 +33,13 @@ export const Hit = ({ hit }) => {
...
       city={city}
       location={location}
-      usRegion={usRegion}
-      intRegion={intRegion}
+      region={region}
       eventUrl={eventUrl}
       subject={subject}
       experience={experience}
       format={format}
-      hit={hit}
```
Comment (ericbakenhus) [inline src/components/events-discovery/components/Hit/Hit.jsx:47]: This looks like a merge that went the wrong way.

Comment (sherakama) [reply src/components/events-discovery/components/Hit/Hit.jsx]: Thanks. I checked out the version from `events-discovery`

**Review by ericbakenhus — APPROVED** [review]
🚀
